EKF/control.cpp

```diff
@@ -899,26 +900,81 @@ void Ekf::controlMagFusion()
 		// or the more accurate 3-axis fusion
 		if (_params.mag_fusion_type == MAG_FUSE_TYPE_AUTO) {
-			// start 3D fusion if in-flight and height has increased sufficiently
-			// to be away from ground magnetic anomalies
-			// don't switch back to heading fusion until we are back on the ground
+			// Check if height has increased sufficiently to be away from ground magnetic anomalies
 			bool height_achieved = (_last_on_ground_posD - _state.pos(2)) > 1.5f;
-			bool use_3D_fusion = _control_status.flags.in_air && (_control_status.flags.mag_3D || height_achieved);
 
-			if (use_3D_fusion && _control_status.flags.tilt_align) {
-				// if transitioning into 3-axis fusion mode, we need to initialise the yaw angle and field states
+			// Check if there has been enough change in horizontal velocity to make yaw observable
+			// Apply hysteresis to check to avoid rapid toggling
+			if (_yaw_angle_observable) {
+				_yaw_angle_observable = _accel_lpf_NE.norm() > _params.mag_acc_gate;
+			} else {
+				_yaw_angle_observable = _accel_lpf_NE.norm() > 2.0f * _params.mag_acc_gate;
+			}
+			_yaw_angle_observable = _yaw_angle_observable && (_control_status.flags.gps || _control_status.flags.ev_pos);
+
+			// check if there is enough yaw rotation to make the mag bias states observable
+			if (!_mag_bias_observable && (fabsf(_yaw_rate_lpf_ef) > _params.mag_yaw_rate_gate)) {
+				// initial yaw motion is detected
+				_mag_bias_observable = true;
+				_yaw_delta_ef = 0.0f;
+				_time_yaw_started = _imu_sample_delayed.time_us;
+			} else if (_mag_bias_observable) {
+				// monitor yaw rotation in 45 deg sections.
+				// a rotation of 45 deg is sufficient to make the mag bias observable
+				if (fabsf(_yaw_delta_ef) > 0.7854f) {
+					_time_yaw_started = _imu_sample_delayed.time_us;
+					_yaw_delta_ef = 0.0f;
+				}
+				// require sustained yaw motion of 50% the initial yaw rate threshold
+				float min_yaw_change_req =  0.5f * _params.mag_yaw_rate_gate * (1e-6f * (float)(_imu_sample_delayed.time_us - _time_yaw_started));
+				_mag_bias_observable = fabsf(_yaw_delta_ef) > min_yaw_change_req;
+			} else {
+				_mag_bias_observable = false;
+			}
+
+			// record the last time that movement was suitable for use of 3-axis magnetometer fusion
+			if (_mag_bias_observable || _yaw_angle_observable) {
+				_time_last_movement = _imu_sample_delayed.time_us;
+			}
+
+			// decide whether 3-axis magnetomer fusion can be used
+			bool use_3D_fusion = _control_status.flags.tilt_align && // Use of 3D fusion requires valid tilt estimates
+					_control_status.flags.in_air && // don't use when on the ground becasue of magnetic anomalies
+					(_flt_mag_align_complete || height_achieved) && // once in-flight field alignment has been performed, ignore relative height
+					((_imu_sample_delayed.time_us - _time_last_movement) < 2000000); // Using 3-axis fusion for a minimum period after to allow for false negatives
```

Comment:
As long as it is interpreted as an integer by the compiler and we don't get a type conversion.